reporter/src/main/kotlin/reporters/AbstractNoticeReporter.kt

```diff
@@ -0,0 +1,164 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package com.here.ort.reporter.reporters
+
+import com.here.ort.model.Identifier
+import com.here.ort.model.LicenseFindingsMap
+import com.here.ort.model.OrtResult
+import com.here.ort.model.config.CopyrightGarbage
+import com.here.ort.model.config.OrtConfiguration
+import com.here.ort.model.licenses.LicenseConfiguration
+import com.here.ort.reporter.LicenseTextProvider
+import com.here.ort.reporter.Reporter
+import com.here.ort.reporter.ResolutionProvider
+import com.here.ort.utils.ScriptRunner
+
+import java.io.OutputStream
+
+abstract class AbstractNoticeReporter : Reporter {
+    companion object {
+        const val NOTICE_SEPARATOR = "\n----\n\n"
+    }
+
+    data class NoticeReport(
+        val headers: List<String>,
+        val findings: Map<Identifier, LicenseFindingsMap>,
+        val footers: List<String>
+    )
+
+    class PreProcessor(
+        ortResult: OrtResult,
+        noticeReport: NoticeReport,
+        copyrightGarbage: CopyrightGarbage,
+        licenseConfiguration: LicenseConfiguration
+    ) : ScriptRunner() {
+        override val preface = """
+            import com.here.ort.model.*
+            import com.here.ort.model.config.*
+            import com.here.ort.model.licenses.*
+            import com.here.ort.spdx.*
+            import com.here.ort.utils.*
+            import com.here.ort.reporter.reporters.AbstractNoticeReporter.NoticeReport
+
+            import java.util.*
+
+            var headers = noticeReport.headers
+            var findings = noticeReport.findings
+            var footers = noticeReport.footers
+
+        """.trimIndent()
+
+        override val postface = """
+
+            // Output:
+            NoticeReport(headers, findings, footers)
+        """.trimIndent()
+
+        init {
+            engine.put("ortResult", ortResult)
+            engine.put("noticeReport", noticeReport)
+            engine.put("copyrightGarbage", copyrightGarbage)
+            engine.put("licenseConfiguration", licenseConfiguration)
+        }
+
+        override fun run(script: String): NoticeReport = super.run(script) as NoticeReport
+    }
+
+    abstract class NoticeProcessor(
+        protected val ortResult: OrtResult,
+        protected val ortConfig: OrtConfiguration,
+        protected val resolutionProvider: ResolutionProvider,
+        protected val licenseTextProvider: LicenseTextProvider,
+        protected val copyrightGarbage: CopyrightGarbage,
+        protected val licenseConfiguration: LicenseConfiguration
+    ) {
+        abstract fun process(noticeReport: AbstractNoticeReporter.NoticeReport): List<() -> String>
```

Comment:
Notice files can get large, and most of the content are license texts. Loading all of them into memory can easily be avoided by using lambdas that produce strings on demand instead of just strings.